Overhaul Docking Module

rbharath opened the pull request:
This PR overhauls the `dc.dock` module. I still need to do some cleanup on it, but splitting it out to make development cleaner.

rbharath commented:
This PR should now be ready to review. It overhauls the `dc.dock` module to be more tightly focused and functional. I ended up adding a number of new utilities functions which helped me implement the docking support features. As a couple of extra notes:

- For `VinaPoseGenerator`, I couldn't figure out how to install Autodock Vina on windows. The autodock vina site provides a `.msi` file. For now, I'm raising an error if this class is used on Windows.
- I had to comment out one test in `test_atomic_coordinates.py`. I'll uncomment this test in my next PR that brings in changes to the atomic coordinate featurizer.

CC @peastman @seyonechithrananda @vsomnath @nd-02110114

peastman commented on deepchem/dock/docking.py:
```diff
@@ -1,119 +1,96 @@
+           num_modes=9,
+           num_pockets=None,
+           out_dir=None):
+    """Docks using Vina and RF.
```
This docstring has a bunch of references to Vina and RF that need to be changed.

rbharath replied:
Good catch. Fixed

peastman commented on deepchem/dock/docking.py:
```diff
@@ -1,119 +1,96 @@
+        # TODO: How to handle the failure here?
+        features, _ = self.featurizer.featurize_complexes([molecular_complex])
+        dataset = NumpyDataset(X=features)
+        score = self.model.predict(dataset)
```
Should this be `self.scoring_model`?

rbharath replied:
Good catch! This code turned out to be buggy. I've added an extra unit test for this path to make sure it's run in the tests.

peastman commented on deepchem/dock/pose_generation.py:
```diff
@@ -1,159 +1,277 @@
+      Of shape `(3,)` holding the size of the box to dock. If not
+      specified is set to size of molecular complex plus 5 angstroms.
+    exhaustiveness: int, optional (default 10)
+      Tells Autodock Vina how exhaustive it should be with pose
+      generation.
+    num_modes: int, optional (default 9)
+      Tells Autodock Vina how many binding modes it should generate at
+      each invocation.
```
Incorrect references to Vina.

rbharath replied:
Good catch, fixed.

peastman commented on deepchem/dock/tests/test_pose_scoring.py:
```diff
@@ -26,34 +25,51 @@ class TestPoseScoring(unittest.TestCase):
+    Nrot = 5
+    w = 0.5
+    out_tensor = vina_nonlinearity(c, w, Nrot)
+    assert out_tensor.shape == (N, M)
```
Just checking the shape isn't much of a test.  Can you verify that the result is actually correct?

rbharath replied:
Added better tests here.

peastman commented on deepchem/feat/binding_pocket_features.py:
```diff
@@ -1,18 +1,61 @@
   """
-  Featurizes binding pockets with information about chemical environments.
+  mapping = {}
+  for box_ind, box in enumerate(boxes):
+    box_atoms = []
+    for atom_ind in range(len(coords)):
+      atom = coords[atom_ind]
+      if atom in box:
+        box_atoms.append(atom_ind)
+    mapping[box] = box_atoms
```
If the number of atoms is large, and if this turns out to be a bottleneck, there are much faster ways of doing it.

rbharath replied:
Would you have suggestions for better algorithms here? It would be useful to have this technique be fast in cast we want to use it at large scale.

peastman replied:
This code scales as O(n*m) where n is the number of atoms and m is the number of boxes.  If n and m are both large, a faster approach is to first assign atoms to voxels, which takes a single loop of O(n), and then for each box loop over the voxels that intersect it, which scales as O(m).

It's probably best to leave this code as it is for the moment, then see how it works in practice and whether it needs to be optimized.

peastman commented on deepchem/utils/pdbqt_utils.py:
```diff
@@ -0,0 +1,315 @@
+  lines = [x[:66] for x in lines]
+  pdb_map = {}
+  for line in lines:
+    my_values = line.split()
```
PDB files are based on fixed column positions.  They are not whitespace delimited.  See http://www.wwpdb.org/documentation/file-format-content/format33/sect9.html#ATOM.

rbharath replied:
Good catch! Fixed this to use the correct column numbers.

peastman commented:
That's a lot of code!  I made a quick pass through it, but it wasn't at all thorough.

Once this is merged I'll see if I can get it to work on Windows.

You've hardcoded the assumption that binding pockets will always be represented by axis aligned bounding boxes.  Is that a good assumption, or do some algorithms use other representations such as spheres?

rbharath commented:
@peastman Thanks for the review! This is a large chunk of code so I very much appreciate the effort :)

That's a great point about the hard coded binding pocket assumption. I think that this assumption is a bit too restrictive. At the least we probably want to allow for non-axis-aligned binding boxes. I haven't seen someone use a spherical binding region yet, but it's definitely not out of the realm of possibility. Let me think a bit about how to clean this up.

I'll do a pass to address the review comments and update this PR early next week

seyonechithrananda commented:
Hey @rbharath! Thanks for tagging me on this PR. I'll take a look this weekend!

rbharath commented:
I've addressed the review comments and I've also added a documentation page for Docking. I also fixed https://github.com/deepchem/deepchem/issues/1885 since it was a small edit.

The only thing this PR hasn't fixed is the hard-coded assumption that binding pockets are always axis aligned boxes. I think this requires some refactoring and I think this might be better done in a separate PR. I'll create an issue to track.

If Travis passes, I think this PR is good to merge.

rbharath commented:
There are two tests that are failing consistently across linux/mac/windows, one in `test_rdkit_grid_featurizer.py` and `test_maml.py`. It's a little confusing since I'm fairly sure this PR doesn't change anything that could cause those tests to fail. I'll dig into this a bit more

rbharath commented:
Travis is green so going to merge this in
